### PR TITLE
Improve exceptions

### PR DESCRIPTION
**Pull request by AlexMasterov**
Error messages are now more informative.
Use a separate file for exceptions improves readability and re-use, and allows to implement [marker (type) exception](https://github.com/AlexMasterov/framework/blob/c55618ae8071a733edaeedb53c93753109b5490e/src/Exception/ExceptionInterface.php) for [exceptions coming from the Equip](https://github.com/AlexMasterov/framework/blob/5be6b87cd3ab85eb78a08e40b523cf82e5c58112/src/Handler/ExceptionHandler.php#L79).

Ref: #56

### CODE REVIEW

File: src/Middleware/MiddlewareSet.php [hunk 1]
```diff
@@ -2,25 +2,24 @@
...
 use Equip\Structure\Set;
 
 class MiddlewareSet extends Set
 {
     /**
      * @inheritDoc
      *
-     * @throws \DomainException if $middlewares does not conform to type expectations
+     * @throws MiddlewareException
+     * If $classes does not conform to type expectations.
```
Comment (shadowhand) [inline src/Middleware/MiddlewareSet.php]: Formatting is off here... it should be in the form:

```php
/**
 * @throws Exception Description.
 */
```

or

```php
/**
 * @throws Exception
 *  Description, note the space before the description starts.
 */

File: src/Responder/FormattedResponder.php [hunk 2]
```diff
@@ -49,8 +49,7 @@ public function __invoke(
         PayloadInterface $payload
     ) {
         if ($this->hasOutput($payload)) {
-            $formatter = $this->formatter($request);
-            $response = $this->format($response, $formatter, $payload);
```
Comment (shadowhand) [inline src/Responder/FormattedResponder.php]: I'm not a fan of mixing this change in with adding exceptions. This PR should contain changes to exceptions **only** and other refactoring should be done in a separate PR.

**Comment by shadowhand** [comment]
✋

**Comment by shadowhand** [comment]
👎

**Comment by AlexMasterov** [comment]
@shadowhand please check now.

File: src/Responder/FormattedResponder.php (hunk 2 above)
Comment (AlexMasterov) [inline src/Responder/FormattedResponder.php]: @shadowhand I agree with you, I will. Just too many changes intersect. 

I would like to hear something about the changes.. (is it worth it?)

File: src/Responder/ChainedResponder.php [hunk 3]
```diff
@@ -51,19 +51,16 @@ public function __invoke(
     /**
      * @inheritDoc
      *
-     * @throws InvalidArgumentException If a responder does not implement the correct interface.
+     * @throws ResponderException
+     *  If $classes does not implement the correct interface.
      */
-    protected function assertValid(array $data)
```
Comment (shadowhand) [inline src/Responder/ChainedResponder.php]: These were called `$data` to match the `Set` parameter name. But that's an okay change.

**Comment by shadowhand** [comment]
Manually merged to `master`, since there are no functional changes here as the new exceptions are extensions of the old type.

**Comment by AlexMasterov** [comment]
@shadowhand Hmm. I'll fix the tests.
